grpc_unary_attestation/src/server.rs

```diff
@@ -18,20 +18,88 @@
 //! protocol.
 
 use crate::proto::{unary_session_server::UnarySession, UnaryRequest, UnaryResponse};
+use lru::LruCache;
+use oak_remote_attestation::handshaker::{AttestationBehavior, Encryptor, ServerHandshaker};
 use oak_utils::LogError;
+use std::sync::{Arc, Mutex};
 use tonic;
 
-/// gRPC Attestation Service implementation.
-pub struct AttestationServer<F, L: LogError> {
+type SessionId = u64;
+enum SessionState {
+    // Boxed due to large size difference, ref: https://rust-lang.github.io/rust-clippy/master/index.html#large_enum_variant
+    HandshakeInProgress(Box<ServerHandshaker>),
+    EncryptedMessageExchange(Encryptor),
+}
+
+/// Maintains remote attestation state for a number of sessions
+struct SessionsTracker {
     /// PEM encoded X.509 certificate that signs TEE firmware key.
-    _tee_certificate: Vec<u8>,
-    /// Processes data from client requests and creates responses.
-    _request_handler: F,
+    tee_certificate: Vec<u8>,
     /// Configuration information to provide to the client for the attestation step.
-    _additional_info: Vec<u8>,
+    additional_info: Vec<u8>,
+    known_sessions: LruCache<SessionId, SessionState>,
+}
+
+impl SessionsTracker {
+    pub fn create(tee_certificate: Vec<u8>, additional_info: Vec<u8>) -> Self {
+        Self {
+            tee_certificate,
+            additional_info,
+            known_sessions: LruCache::new(10000),
+        }
+    }
+
+    /// Consumes remote attestation state of an existing session. Creates
+    /// intial state if the session is not known.
+    ///
+    /// Note that getting the remote attestaion state of a session always
+    /// implicitly removes it from the set of tracked sessions. After
+    /// using the state to process a request with this state it must explicitly
+    /// be put back into the SessionsTracker. This an intentional choice meant
+    /// meant to ensure that faulty state that leads to errors when preocessing
+    /// a request is not persistent.
+    pub fn pop_session_state(&mut self, session_id: SessionId) -> Result<SessionState, String> {
+        return match self.known_sessions.pop(&session_id) {
```

Comment:
it's a tiny function that barely wraps the pop operation. cloning additional_info is the most computationally it does. But agreed that there's a possibility of scope creep with future changes that might be worth guarding against. 

If we move the Mutex into `SessionTracker` (while keeping it behind an Arc) we'll no longer be able to get the returned session_state as mutable on L149. What do you think is the best way to solve that?

Two solutions I can think of:

- We could clone the state, but it means we'd have to implement cloning for all the various structs used in the handshaker / encryptor. Not sure we want that.
- We could merge `SessionsTracker` and `AttestationServer`, extracting the normalization logic currently in `pop_session_state` into a static function. But it wouldn't be as clean of a separation of concerns. IMO if we expose the native LRU instance there's a significant the risk of future changes leading to inconsistently applied normalization.

Reply:
I don't see why it would not be able to be mutable. What is the compiler error? This removes it from the LRU cache, so there should no longer be anything that has a reference to it once it is returned from this function, so you should be able to make it mutable.

Reply:
I also would have thought that it's sufficient to limit the lock here, given the entry is removed (popped) from the LRU cache, so there is no need to hold references or clone. Or am I not understanding what pop actually does?

Reply:
See the compiler error, given on L149 & L199 with the inner mutex:

> "cannot borrow data in an `Arc` as mutable
> trait `DerefMut` is required to modify through a dereference, but it is not implemented for `std::sync::Arc<SessionsTracker>`rustc[E0596](https://doc.rust-lang.org/error-index.html#E0596)" 

For convince of debugging, I've also pushed the resulting code onto the following branch which you can clone from my fork: `unary-handler-inner-mutex`

Reply:
I haven't tried that code, but the problem may be that the Mutex and Arc are "too far" from each other. The usual idiom is to have `Arc<Mutex<Thing>>` (like you have here). Also in that code I'm not sure I understand why the entire SessionTracker is inside an Arc -- shouldn't there be a single instance of it ever? Who / what creates other references to it? Anyways, if that's needed, then maybe an additional Arc may be needed to wrap the Mutex anyways? Not sure though

Reply:
Investigated this a little. I think what makes this behavior a bit hard to reason about for me is the `#[tonic::async_trait]` (which is [this crate](https://github.com/dtolnay/async-trait) under the hood), which seems a little magical (or at least I'm not sure what it does precisely). Do you know more about it?

There is indeed only ever one instance of the `SessionTracker` per `AttestationServer`, and you're correct that we indeed do not need an Arc. 

Invoking the async `message` function in my understanding does start a new thread, hence why I assumed an Arc was needed. It isn't though, I assume `#[tonic::async_trait]` already provides each thread with a shared reference to the `self` object. 

Adding an Arc inside `SessionTracker` does not fix it. It also seems like it wouldn't be needed, there's only ever one session_tracker.

Reply:
I would be inclined to move forward with this PR with the outer mutex for now. However I'd be keen to debug this further later, would like to understand what exactly is needed.

Reply:
SG, thanks